grails-app/services/de/iteratec/osm/system/MissingJobResultCheckService.groovy

```diff
@@ -0,0 +1,103 @@
+
+    JobResultPersisterService jobResultPersisterService
+
+    def fillMissingJobResults() {
+        Date latestCheckDate = getLatestCheck()?.date
+        int missingResults = 0
+        if(latestCheckDate) {
+            missingResults = fillMissingJobResultsSinceDate( new DateTime(latestCheckDate).minusMinutes(SCHEDULE_THRESHOLD_MINUTES.toInteger()))
+            new MissingJobResultCheck(date: new Date(), missingResults: missingResults).save()
+        }
+        else {
+            DateTime dateDaysAgo =  new DateTime().minusDays(FIRST_CHECK_DAYS_AGO)
+            missingResults = fillMissingJobResultsSinceDate( dateDaysAgo )
+            new MissingJobResultCheck(date: new Date(), missingResults: missingResults).save()
+        }
+        log.info("Fixed Missing JobResults: $missingResults")
+    }
+
+    private int fillMissingJobResultsSinceDate(DateTime since) {
+        int missingJobResults = 0
+        getPreparedJobList(since).forEach{ job ->
+            def existingResults = getJobResultsSince(job.job, job.since)
+            missingJobResults += checkForMissingJobResults(existingResults, job.since, job.job)
+        }
+        return missingJobResults
+    }
+
+    private List<Map> getPreparedJobList(DateTime since) {
+        return Job.findAllByDeletedAndActiveAndExecutionScheduleIsNotNull(false, true).collect {
+            ['job': it, 'id': it.id, 'since': since]
+        }
+    }
+
+    private def getJobResultsSince(Job job, DateTime since) {
+        return JobResult.findAllByJobAndDateGreaterThan(job, since.toDate())
+    }
+
+    private int checkForMissingJobResults( List<JobResult> existingResults, DateTime since, Job job) {
+        DateTime now = new DateTime()
+        List<DateTime> scheduledDates = getScheduledDates(since.toDate(), now.toDate(), job)
+        existingResults.forEach{ jobResult ->
+            if(jobResult.executionDate && (jobResult.executionDate <= now.toDate())) {
+                DateTime jobResultDate = findDateForJobResult(scheduledDates, jobResult )
+                if(jobResultDate) {
+                    scheduledDates.remove(jobResultDate)
+                }
+            }
+        }
+        if(!scheduledDates.empty) {
+            log.info("Persist ${scheduledDates.size()} Missing JobResults for: $job")
+            persistMissingJobResults(scheduledDates, job)
+        }
+        return scheduledDates.size()
+    }
+
+    DateTime findDateForJobResult(List<DateTime> dates, JobResult jobResult) {
+        DateTime execDate = new DateTime(jobResult.executionDate)
+        List<DateTime> closeDates = dates.findAll { Math.abs(dateDist(it, execDate)) < SCHEDULE_THRESHOLD_MINUTES * (1000L * 60L) }
+        if(!closeDates.empty) { // Find closest date
+            return (closeDates.collect{ [Math.abs(dateDist(it, execDate)), it] }.sort().first())[1]
+        }
+        return null
+    }
+
+    long dateDist(DateTime d1, DateTime d2) {
+        return Math.abs((d1 - d2.millis).millis)
```

Comment:
This code looks actually quite confusion, since you first substract milliseconds from a `DateTime` and try to access the millis of the result.
`Math.abs(d1.millis - d2.millis)` would be more efficient (doesn't create a new object) and is easier to understand.